beluga_tutorial/src/main.cpp

```diff
@@ -0,0 +1,278 @@
+    tutorial_params.dt = params["dt"].as<double>();
+    tutorial_params.velocity = params["velocity"].as<double>();
+    tutorial_params.translation_sd = params["translation_sd"].as<double>();
+    tutorial_params.sensor_range = params["sensor_range"].as<double>();
+    tutorial_params.sensor_model_sigam = params["sensor_model_sigam"].as<double>();
+    tutorial_params.min_particle_weight = params["min_particle_weight"].as<double>();
+    tutorial_params.dataset_path = params["dataset_path"].as<std::string>();
+    tutorial_params.landmark_map_file_name = params["landmark_map_file_name"].as<std::string>();
+  } catch (YAML::BadFile& e) {
+    std::cout << e.what() << "\n";
+  }
+}
+
+void load_landmark_map_from_yaml(const std::filesystem::path& path, LandmarkMapVector& landmark_map) {
+  try {
+    YAML::Node node = YAML::LoadFile(path);
+    landmark_map = node["landmark_map"].as<LandmarkMapVector>();
+  } catch (YAML::BadFile& e) {
+    std::cout << e.what() << "\n";
+  }
+}
+
+void save_tutorial_dataset_to_yaml(const std::filesystem::path& path, const TutorialDataset& tutorial_dataset) {
+  YAML::Node node;
+
+  for (const auto& lmp : tutorial_dataset.landmark_map) {
+    node["landmark_map"].push_back(lmp);
+  }
+
+  for (auto&& [cycle, data] : tutorial_dataset.sim_data | ranges::views::enumerate) {
+    node["simulation_dataset"][cycle]["ground_truth"] = data.ground_truth;
+
+    auto current = node["simulation_dataset"][cycle]["particles"]["current"];
+    current["states"] = beluga::views::states(data.current) | ranges::to<std::vector<double>>;
+    current["weights"] = beluga::views::states(data.current) | ranges::to<std::vector<double>>;
+
+    auto propagate = node["simulation_dataset"][cycle]["particles"]["propagate"];
+    propagate["states"] = beluga::views::states(data.propagate) | ranges::to<std::vector<double>>;
+    propagate["weights"] = beluga::views::states(data.propagate) | ranges::to<std::vector<double>>;
+
+    auto reweight = node["simulation_dataset"][cycle]["particles"]["reweight"];
+    reweight["states"] = beluga::views::states(data.reweight) | ranges::to<std::vector<double>>;
+    reweight["weights"] = beluga::views::states(data.reweight) | ranges::to<std::vector<double>>;
+
+    auto resample = node["simulation_dataset"][cycle]["particles"]["resample"];
+    resample["states"] = beluga::views::states(data.resample) | ranges::to<std::vector<double>>;
+    resample["weights"] = beluga::views::states(data.resample) | ranges::to<std::vector<double>>;
+
+    auto estimation = node["simulation_dataset"][cycle]["estimation"];
+    estimation["mean"] = std::get<0>(data.estimation);
+    estimation["sd"] = std::get<1>(data.estimation);
+  }
+
+  std::filesystem::create_directory(path.parent_path());
+  std::ofstream fout(path, std::ios::trunc);
+  fout << node;
+  fout.close();
+}
+
+int main(int argc, char* argv[]) {
+  if (argc < 3) {
+    std::cerr << "Usage: " << argv[0] << "<params_directory_path> <params_filename>\n";
```

Comment:
It is true that this way we save a parameter but at the same time the execution command is longer, I believe that from the parameter it can be more comfortable but it is a question of user's experience.

Reply:
In that case, consider taking a single path parameter, instead of the directory and the file name.

Reply:
I also thought about that, but I need the directory to load the map. Either I move the map to the param file, or I create another path for the map and put this path in the param file.

Reply:
Can map be part of the param yaml file? So we have a single input.

Reply:
Yes it can be, I'll move the map to the param file